lib/Pipeline/Runner.cpp

```diff
@@ -494,3 +506,18 @@ llvm::Error Runner::apply(const GlobalTupleTreeDiff &Diff,
 
   return invalidate(Map);
 }
+
+llvm::Error Runner::purge(const revng::DirectoryPath &ExecutionDirectory) {
```

Comment:
Move this function to `PipelineManager`, you need to `StorageClient.commit()`

Reply:
Previously this cleanup logic did not include a `StorageClient.commit()`, [reference](https://github.com/khaledismaeel/revng/blob/8cd571335b56b0910af3272bd114cc939392479e/lib/Pipes/PipelineManager.cpp#L141). Was that a bug?

Reply:
Ah yes, it was a bug.

Reply:
Oh or actually not... It would __eventually__ be committed [here](https://github.com/revng/revng/blob/8cd571335b56b0910af3272bd114cc939392479e/lib/Pipes/PipelineManager.cpp#L258), the changes were not lost then, and they're not lost now.

We can keep the commit out of `purge`, not to make a redundant commit call, which in my understanding is undesirable.

Reply:
You're right, `commit` gets eventually called.
I'd still move it inside `PipelineManager`, since it already handles most of the storage logic.

Reply:
Looking at one (of the two) uses of `purge` [here](https://github.com/khaledismaeel/revng/blob/c7e33dd26041caf173add660d918343aec95386c/lib/Pipes/PipelineManager.cpp#L288), `purge` is being called within `setUpPipeline` (both directly and in `checkComponentsVersion`), before the pipeline is finalized. So even though `PipelineManager` handles most of the storage logic, this particular logic seems needed in the `Runner`.

I don't think I can move `purge` into `PipelineManager` without some significant refactorings of that portion of the code. Do you still suggest to go ahead and move it?

Reply:
> I don't think I can move purge into PipelineManager without some significant refactorings

What do you mean? To me it seems a pretty simple cut-paste matter. Is there something I'm missing?
Both uses are after `MaybePipeline` has the actual runner. I'd:
1. Move the assignment of `PipelineManager.Runner` to just after the `return MaybePipeline.takeError()`
2. Cut the `purge` method out of `Runner` and into `PipelineManager` as a private method
3. Add `Runner->` where needed

Reply:
That is doable and simple, but I am unsure about setting `PipelineManager.Runner` before being certain all initialization on it has returned without an error. If you think this is okay to do, I'll proceed with it.

Reply:
If you're worried, you can always add `revng_assert(Runner != nullptr)` at the beginning of `purge`
However this is still part of `PipelineManager::create*` so the actual object is never returned to the caller

Reply:
> If you're worried, you can always add revng_assert(Runner != nullptr) at the beginning of purge

That's not what I was worried about; I was worried that setting `PipelineManager.Runner` at the beginning of `setUpPipeline` (and not the end) might leave the pipeline manager in an inconsistent state in case any of the operations on `MaybePipeline` returns with an error.

But maybe I'm just being paranoid; seeing how calm you are about these changes, I'm not worried anymore :D I will implement your suggestion now.

Reply:
Refactored now. I took the liberty to move `checkComponentsVersion` to be a private method on `PipelineManager` as well.

Let me know if the changes are okay.

Reply:
I don't see it in the commits, but sure, go ahead!